### PR TITLE
[cpu] Integrate IStaticShapeInfer wirth IShapeInfer

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - The `IStaticShapeInfer` interface extends `IShapeInfer`.
 - Remove `NgraphShapeInfer` class as its functionality is replaced by `IStaticShapeInfer`.
 - Refactor shape inference unit test to avoid names clashes with CPU plugin types:
   - use `ov::Shape` to avoid interpretation as `intel_cpu::Shape`.
   - rename test type `ShapeVector` to `StaticShapeVector`.

### Tickets:
 - CVS-118704

### CODE REVIEW

File: src/plugins/intel_cpu/src/nodes/interpolate.cpp [hunk 1]
```diff
@@ -1763,27 +1762,21 @@ class InterpolateShapeInferFactory : public ShapeInferFactory {
...
-                port_mask = PortMask(Interpolate::SCALES_ID, Interpolate::AXES_ID);
-            } else if (attr.shape_calculation_mode == ngInterpShapeCalcMode::SIZES) {
-                port_mask = PortMask(Interpolate::TARGET_SHAPE_ID, Interpolate::AXES_ID);
-            } else {
-                OPENVINO_ASSERT(false, "Unsupported interpolate shape calculation mode");
-            }
+            const auto is_supported_mode = (attr.shape_calculation_mode == ngInterpShapeCalcMode::SCALES) ||
+                                           (attr.shape_calculation_mode == ngInterpShapeCalcMode::SIZES);
+            OPENVINO_ASSERT(is_supported_mode, "Unsupported interpolate shape calculation mode");
+            return make_shape_inference(m_op);
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/plugins/intel_cpu/src/nodes/interpolate.cpp:1770]: I'm not sure if custom factory is required as is CPU, this is reason why left to avoid regressions.
I follow your suggestion if it can be safely removed. I Can do it as separate PR to not block working part.

File: src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp [hunk 2]
```diff
@@ -582,18 +612,50 @@ const IStaticShapeInferFactory::TRegistry IStaticShapeInferFactory::registry{
 #undef _OV_OP_SHAPE_INFER_MASK_REG
 #undef _OV_OP_SHAPE_INFER_VA_REG
 
+class ShapeInferCustomMask : public IShapeInfer {
+public:
+    ShapeInferCustomMask(ShapeInferPtr shape_infer, port_mask_t port_mask)
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp:617]: I agree with approach that shape inference should provide.
Its done to fit current usage for custom CPU shape inference which defines mask depends on runtime properties of operator. The shape inference provide compile time value.
The affected cpu nodes:
 - eye.cpp
 - deconv.cpp
 - reference.cpp

If you can confirm is not required in this implementations I will remove it. Or I can make as separate PR to see potential issues in tests. Will follow your recommendation.

File: src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp [hunk 3]
```diff
@@ -151,6 +151,36 @@ class ShapeInferBase : public IStaticShapeInfer {
...
+            result.dims.reserve(shape_infer_result->size());
+            std::transform(shape_infer_result->begin(),
+                           shape_infer_result->end(),
+                           std::back_inserter(result.dims),
+                           [](StaticShape& s) {
+                               return std::move(*s);
+                           });
+        }
+
+        return result;
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp]: Yes, less is better but I'd like also avoid temporary variables which even move has to be destroyed (can be optimized or is meaningless).

Changed to have on switch and no temporary dims to move.

File: src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxnick) [reply src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp:613]: > The affected cpu nodes:
eye.cpp
deconv.cpp
reference.cpp

1. eye.cpp - the custom factory may be removed, as it looks like the only factor defining the mask is the number of inputs. But in the reality there may be the only mask defining all the possible inputs the shape infer depends on, and if the 4th input is absent, it's not a problem as it won't be considered anyway as the routines analyzing the inputs dependency and preparing the data consider the number of inputs.
2. deconv.cpp - indeed tricky part, as the deconvolution layer may fuse bias and it becomes the 3rd or 4th input depending on the optional input of the ngraph operation. But, in such scenarios we would probably have to simply write a wrapper over the default shape infer in order to modify the behavior considering plugin specifics, rather than keep the custom masks in the generic shape inference part.
3.  reference.cpp this is even more interesting case, as there may be any arbitrary type of nod (internaly and externaly dynamic), if the op is provided by the user. But if it's an operation from opset, we could implement the existing port mask for the operation. But it will require some modifications of the reference node implementation itself.

So to summarize. We can remove custom shape inference factory for eye. For deconv and reference, it looks more logical to write a simple wrapper class for shape inference, rather than keep custom mask allowed for the generic shape inference mechanism. What do you think?

[the remark this answers is not in the document]
Comment (praasz) [reply src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp:613]: If this PR pass all check then I will make another one where try:
1. Remove custom shape inference factory for eye, deconv, reference.
2. If there will be any issues with deconv and reference then simple wrappers can be introduced. Or depends on errors reason other solution can be applied.

[the remark this answers is not in the document]
Comment (maxnick) [reply src/plugins/intel_cpu/src/shape_inference/shape_inference.cpp:613]: Ok. Let's agree to remove custom mask in a follow up PR.

**Review by maxnick — APPROVED** [review]
